### ISSUE TITLE
StreamRendering partial DOM Preservation

### ISSUE THREAD

**Issue by garrettlondon1** [dotnet/aspnetcore#57358]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I am trying to conditionally streamrender parts of a page, and preserve the DOM of other parts until the fetch request is completed

### Describe the solution you'd like

I want to be able to stream render only parts of a blazor ssr page

**Comment by javiercn** [comment]
@garrettlondon1 thanks for contacting us.

That's unfortunately not possible by the nature of attributes, as they take in constant-like values that are resolved at compile/initialization time and can't be changed afterwards.

What in particular are you trying to achieve? What are you trying to check on the HTTP Context to decide.

**Comment by garrettlondon1** [comment]
I am trying to stream render a page of tabs

Imagine the tabs are dependent on async data to finish completing

When stream rendering, the tabs will flash because the data is not loaded and you cannot show the tabs if you are not confident the user can see them

Without stream rendering, the DOM is not updated until response is received from the fetch request


I don't need to conditionally stream render, but rather, choose which elements on the page DOM swap before the response and after the response

If I can persist the tab structure on stream rendering, but am allowed to stream render other elements on the page. It will allow a SPA feel to the tab structure where it does not render an empty page before the tab logic loads

Currently: when you click on a tab without stream rendering, and the tab is wrapped around some condition waiting on async data..
- The tab structure will disappear and reappear due to DOM swapping of stream rendering




Maybe we should change this to: Persist certain elements in the DOM while stream rendering, until end of response is received

**Comment by javiercn** [comment]
@garrettlondon1 thanks for the details.

Are we talking about a pure SSR scenario or about the tabs flashing when they become interactive? 

Or are we talking about the tab structure disappearing on the next enhanced navigation?

**Comment by garrettlondon1** [comment]
> @garrettlondon1 thanks for the details.
> 
> Are we talking about a pure SSR scenario or about the tabs flashing when they become interactive?
> 
> Or are we talking about the tab structure disappearing on the next enhanced navigation?

Tab structure disappearing on the next enhanced navigation. That is the biggest issue, because it's wrapped in conditions that are async

Preserving conditional NavLinks in streamrendered scenarios just makes sense, tabs or not

**Comment by garrettlondon1** [comment]
https://github.com/user-attachments/assets/92f7831b-33e3-44b4-9a57-9b2bc94970af

Example of an if condition around the Third tab in the group

**Comment by iphdav** [comment]
@garrettlondon1 Just a thought:  If this is about permissions (which users should see what tab), have you considered either putting those permissions in a cookie (in a similar way to when roles/groups can be included in a JWT).  Or have you thought about caching the information you are looking up?  Or putting it into some type of context, such as SessionState (for StreamRendering where HttpContext is available) or even some service injected with "scoped" dependency (for interactive server rendering).  Would not that simply avoid this issue?

**Comment by iphdav** [comment]
@garrettlondon1 Maybe a few other tricks:  You could emit the data-permanent tag after the first render (that includes the actual menu items you want displayed), which if I understand correctly would stop that being updated from then on unless you do a navigation with the pageReload option set to true (which you could do after some security settings change to what tabs the user should see):
```razor
<div data-permanent>
    ...
</div>
```
Or course, another option, you can always make that an Interactive Server component, and then basically solve your issue any number of ways.

**Comment by iphdav** [comment]
@garrettlondon1 Here is a solution if you want to use data-permanent and streaming rendering.  It does basically implement what you are asking for, which is when using enhanced rendering to delay updating the flickering component until the end of your ```OnInitializedAsync()``` method.  I actually *have* tested this and it does what you need.

First wrap the content of the flickering component in a div like this:
```razor
<div data-permanent="@dataPermanent">
    ...your content
</div>
```
Create a field in your Razor Component like this:
```razor
bool dataPermanent = true;
```
Then at the very last bottom lines of your OnInitializeAsync method include these lines of code:
```razor
dataPermanent = false;
StateHasChanged();
await Task.Yield();
dataPermanent = true;
```
If you finish OnInitializeAsync with those lines, it will effectively do exactly what you are asking.  I would provide a further explanation as to why, but I am 100% confident you will figure it out yourself ;-)

**Comment by iphdav** [comment]
@garrettlondon1 Here is a version wrapped in a component.  Lets say you want this behavior in a few places, you could use it like this in one of your pages doing StreamRendering:
```razor
<DataPermanent Loaded="loaded">
  ...rest of your markup here
</DataPermanent>

@code {
    bool loaded = false;
    override protected async Task OnInitializedAsync()
    {
        \\ Your code doing async stuff here
        loaded = true; \\ Last line of code
    }
}
```

The component would be called DataPermanent.razor and be this:
```razor
<div data-permanent="@dataPermanent">
    @ChildContent
</div>
@code {
    bool dataPermanent = true;
    [Parameter] public RenderFragment ChildContent {get; set;}
    [Parameter] public bool Loaded {get; set;}

    protected override async Task OnParametersSetAsync()
    {
        if (Loaded)
        {
            dataPermanent = false;
            StateHasChanged();
            await Task.Yield();
            dataPermanent = true;
        }
    }
}
```
I hope that helps.

**Comment by garrettlondon1** [comment]
Thanks @iphdav 

The data-permanent solution did kind of work, but it prevents CSS from being applied after the next enhanced navigation, as seen in this issue @SteveSandersonMS created I believe.

The data-permanent says, "on enhanced navigation, keep this element exactly as is"

This story is saying, "on enhanced navigation, don't swap the element until the streaming ends"

https://github.com/dotnet/aspnetcore/issues/51021

**Comment by javiercn** [comment]
@garrettlondon1 I've filed a separate issue https://github.com/dotnet/aspnetcore/issues/57399 to more clearly define a proposal.

If you have the sample with you use case and can put it on a repo that would help. I'm still not 100% sure there aren't different and potentially more straightforward ways of achieving the same outcome. The scenario is a bit unclear to me still, in that I don't yet fully grasp why would you want to stream render on the first place if later on its not acceptable.

If you can provide the code in a repo, that would help.

**Comment by javiercn** [comment]
Tracking the request on https://github.com/dotnet/aspnetcore/issues/57399
